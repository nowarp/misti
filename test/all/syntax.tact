//
// Checks if the tools process various Tact syntax constructions.
//

trait T {
  abstract get fun getter(): Int;
  abstract const something: Int;
}
contract TestContract with T {
  override get fun getter(): Int { return 0 }
  override const something: Int = 42;
}
trait Z {
  a: Int;
  get fun getter(): Int {
      return self.a + 1;
  }
}
contract TestContractF {
    a: Int = 42;
    fun test() {}
}
trait TParent {
  a: Int;
  fun getA(): Int {
      return self.a;
  }
}
trait TChild with TParent {
  a: Int; // Declare 'a' to satisfy TParent's requirement
}
contract TestContractT with TChild {
    a: Int = 19;
    fun test(): Int { return self.getA(); }
    external() {}
}

fun test_try(a: Int) {
    try { a += 1; }
    try { /* empty */ }
    try { a = 19; } catch (err) { dump(err); }
}

fun test_loops() {
    let sum: Int = 0;
    let i: Int = 0;
    while (i < 10) {
      i = i + 1;
      sum = sum + i;
    }
    do {
      i = i - 1;
      sum = sum + i;
    } until (i > 10);
    repeat (10) {
      i = i + 1;
      sum = sum + i;
    }
    let m: map<Int, Int> = emptyMap();
    foreach (k, v in m) {}
}
