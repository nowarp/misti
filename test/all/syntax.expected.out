[INFO] DumpIsUsed: Found `dump` usage
test/all/syntax.tact:41:35:
  40 |     try { /* empty */ }
> 41 |     try { a = 19; } catch (err) { dump(err); }
                                         ^
  42 | }
Help: Using `dump` in production code can sometimes indicate complex code that requires additional review
See: https://nowarp.io/tools/misti/docs/detectors/DumpIsUsed

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:48:7:
  47 |     while (i < 10) {
> 48 |       i = i + 1;
             ^
  49 |       sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:49:7:
  48 |       i = i + 1;
> 49 |       sum = sum + i;
             ^
  50 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i -= 1
test/all/syntax.tact:52:7:
  51 |     do {
> 52 |       i = i - 1;
             ^
  53 |       sum = sum + i;
Help: Consider using augmented assignment instead: i -= 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:53:7:
  52 |       i = i - 1;
> 53 |       sum = sum + i;
             ^
  54 |     } until (i > 10);
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:56:7:
  55 |     repeat (10) {
> 56 |       i = i + 1;
             ^
  57 |       sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:57:7:
  56 |       i = i + 1;
> 57 |       sum = sum + i;
             ^
  58 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign